shell/platform/android/test/io/flutter/embedding/engine/dart/DartMessengerTest.java

```diff
@@ -173,17 +179,63 @@ public void cleansUpMessageData() throws InterruptedException {
     final DartMessenger messenger = new DartMessenger(fakeFlutterJni, () -> synchronousTaskQueue);
-    BinaryMessenger.TaskQueue taskQueue = messenger.makeBackgroundTaskQueue();
-    String channel = "foobar";
+    final BinaryMessenger.TaskQueue taskQueue = messenger.makeBackgroundTaskQueue();
+    final String channel = "foobar";
     BinaryMessenger.BinaryMessageHandler handler =
         (ByteBuffer message, BinaryMessenger.BinaryReply reply) -> {
           throw new RuntimeException("hello");
         };
     messenger.setMessageHandler(channel, handler, taskQueue);
     final ByteBuffer message = ByteBuffer.allocateDirect(4 * 2);
-    int replyId = 1;
-    long messageData = 1234;
+    final int replyId = 1;
+    final long messageData = 1234;
+
     messenger.handleMessageFromDart(channel, message, replyId, messageData);
     verify(fakeFlutterJni).cleanupMessageData(eq(messageData));
   }
+
+  @Test
+  public void emptyResponseWhenHandlerIsNotSet() throws InterruptedException {
+    final FlutterJNI fakeFlutterJni = mock(FlutterJNI.class);
+    final DartMessenger messenger = new DartMessenger(fakeFlutterJni, () -> synchronousTaskQueue);
+    final String channel = "foobar";
+    final ByteBuffer message = ByteBuffer.allocateDirect(4 * 2);
+    final int replyId = 1;
+    final long messageData = 1234;
+
+    messenger.handleMessageFromDart(channel, message, replyId, messageData);
+    shadowOf(getMainLooper()).idle();
+    verify(fakeFlutterJni).invokePlatformMessageEmptyResponseCallback(replyId);
+  }
+
+  @Test
+  public void buffersResponseWhenHandlerIsNotSet() throws InterruptedException {
+    final FlutterJNI fakeFlutterJni = mock(FlutterJNI.class);
+    final DartMessenger messenger = new DartMessenger(fakeFlutterJni, () -> synchronousTaskQueue);
+    final BinaryMessenger.TaskQueue taskQueue = messenger.makeBackgroundTaskQueue();
+    final String channel = "foobar";
+    final ByteBuffer message = ByteBuffer.allocateDirect(4 * 2);
+    final int replyId = 1;
+    final long messageData = 1234;
+
+    messenger.enableBufferingIncomingMessages();
+    messenger.handleMessageFromDart(channel, message, replyId, messageData);
+
+    shadowOf(getMainLooper()).idle();
+    verify(fakeFlutterJni, never()).invokePlatformMessageEmptyResponseCallback(eq(replyId));
+
+    final BinaryMessenger.BinaryMessageHandler handler =
+        (ByteBuffer msg, BinaryMessenger.BinaryReply reply) -> {
+          reply.reply(ByteBuffer.wrap("done".getBytes()));
+        };
+    messenger.setMessageHandler(channel, handler, taskQueue);
+
+    shadowOf(getMainLooper()).idle();
+    verify(fakeFlutterJni, never()).invokePlatformMessageEmptyResponseCallback(eq(replyId));
+
+    final ArgumentCaptor<ByteBuffer> response = ArgumentCaptor.forClass(ByteBuffer.class);
+    verify(fakeFlutterJni)
+        .invokePlatformMessageResponseCallback(anyInt(), response.capture(), anyInt());
+    assertArrayEquals("done".getBytes(), response.getValue().array());
+  }
```

Comment:
Can you add a test that handles when a handler gets _unset_ and a message is sent from Dart?

I'm not entirely clear on how we should handle that, but it seems like that should end up throwing a MissingPluginException. But maybe it's not a case we're too concerned about, I can't think of why this would happen.

A case that _is_ a bit more concerning is someone forgets to ever register the plugin but still tries to send a message on it. E.g. when developing a plugin you have a bug you can't really discover easily.

Reply:
Done